added about useRef() topic hook in react (#3858)
Codecademy/docs: content/react/concepts/hooks/terms/useRef/useRef.md

---
Title: 'useState()'
Description: 'Create mutable references to elements or values, allowing access to them without causing re-renders.'
Subjects:
  - 'Web Development'
Tags:
  - 'Components'
  - 'React'
CatalogContent:
  - 'react-101'
  - 'paths/front-end-engineer-career-path'
---

The **`useRef()`** hook in React is used to create mutable references to elements or values within functional [components](https://www.codecademy.com/resources/docs/react/components). It allows you to access and interact with DOM elements directly and persist data across renders without causing re-renders.

## Syntax

In the `useState()` hook, the `initialValue` is an optional parameter that you can provide when creating a new ref object.

`initialValue` allows you to initialize the `.current` property of the ref object with an initial value.

> Here is how you can use this hook :

Reviewer: ```suggestion
This hook can be utilized as follows:
```

```pseudo
import React, { useRef } from 'react';

const myRef = useRef(initialValue);
```

When `useRef()` is called, it returns a ref object, and this ref object has a `.current` property.

> If no `initialValue` is provided, `myRef.current` will be undefined by default.

Reviewer: ```suggestion
If no `initialValue` is provided, `myRef.current` will be null by default.
```

## Example 1

In this example, the useRef hook is used to store the previous value of the count state, allowing you to display the current and previous count values without triggering re-renders.

```jsx
import React, { useState, useRef, useEffect } from 'react';

function PreviousValue() {
  const [count, setCount] = useState(0);
  const prevCountRef = useRef();

  useEffect(() => {
    prevCountRef.current = count;
  });

  return (
    <div>
      <p>Current Count: {count}</p>
      <p>Previous Count: {prevCountRef.current}</p>
      <button onClick={() => setCount(count + 1)}>Increment</button>
    </div>

>   );

Reviewer: Please add an explanation of the code.

Reviewer: Hey @kaushal1717,you forgot to add explanation of the example.

---

> Title: 'useState()'

Reviewer 2: ```suggestion
Title: 'useRef()'
```

> Description: 'Create mutable references to elements or values, allowing access to them without causing re-renders.'

Reviewer 2: ```suggestion
Description: 'Creates mutable references to elements or values, allowing access to them without causing re-renders.'
```

Subjects:
  - 'Web Development'
Tags:
  - 'Components'
  - 'React'
CatalogContent:
  - 'react-101'
  - 'paths/front-end-engineer-career-path'
---

The **`useRef()`** hook in React is used to create mutable references to elements or values within functional [components](https://www.codecademy.com/resources/docs/react/components). It enables direct access and interaction with DOM elements, allowing the persistence of data across renders without triggering re-renders.

## Syntax

> In the `useRef()` hook, the `initialValue` is an optional parameter that you can provide when creating a new ref object.

Reviewer 2: ```suggestion
In the `useRef()` hook, the `initialValue` is an optional parameter that can be provided when creating a new ref object.
```
We avoid first- and second-person pronouns in Docs.

> `initialValue` allows you to initialization of the `.current` property of the ref object with an initial value.

Reviewer 2: ```suggestion
`initialValue` allows the initialization of the `.current` property of the ref object with an initial value.
```

This hook can be utilized as follows :

```pseudo
import React, { useRef } from 'react';

const myRef = useRef(initialValue);
```

When `useRef()` is called, it returns a ref object, and this ref object has a `.current` property.

If no `initialValue` is provided, `myRef.current` will be null by default.

## Example 1

> In this example, the `useRef()` hook is used to store the previous value of the count state, allowing you to display the current and previous count values without triggering re-renders.

Reviewer 2: ```suggestion
In this example, the `useRef()` hook is used to store the previous value of the count state, allowing the display of the current and previous count values without triggering re-renders.
```

```jsx
import React, { useState, useRef, useEffect } from 'react';

function PreviousValue() {
  const [count, setCount] = useState(0);
  const prevCountRef = useRef();

  useEffect(() => {
    prevCountRef.current = count;
  });

  return (
    <div>
      <p>Current Count: {count}</p>
      <p>Previous Count: {prevCountRef.current}</p>
      <button onClick={() => setCount(count + 1)}>Increment</button>
    </div>
  );
}
```

In the code example, useRef is used to create the `prevCountRef` reference to store and access the previous value of the count state variable.

To store the previous count value in a state variable (using `useState()`), each time the `Increment` button is clicked and the count changes, it would trigger a re-render. Because changing state variables causes components to re-render.

> There is no need to re-render the component just to track the previous value; we only need to update the value without causing a re-render. So we use `useRef()` hook

Reviewer 2: ```suggestion
There is no need to re-render the component just to track the previous value; it's only necessary to update the value without causing a re-render. So the `useRef()` hook is used.
```
